cpp/velox/shuffle/VeloxShuffleWriter.cc

```diff
@@ -167,8 +167,78 @@ std::shared_ptr<arrow::Array> makeBinaryArray(
   return arrow::MakeArray(arrow::ArrayData::Make(type, 1, {nullptr, offsetBuffer, valueBuffer}));
 }
 
+inline void writeInt64(std::shared_ptr<arrow::Buffer> buffer, int64_t& offset, int64_t value) {
+  memcpy(buffer->mutable_data() + offset, &value, sizeof(int64_t));
+  offset += sizeof(int64_t);
+}
+
+int64_t getMaxCompressedBufferSize(
+    const std::vector<std::shared_ptr<arrow::Buffer>>& buffers,
+    arrow::util::Codec* codec) {
+  int64_t totalSize = 0;
+  for (auto& buffer : buffers) {
+    if (buffer != nullptr) {
+      totalSize += codec->MaxCompressedLen(buffer->size(), nullptr);
+    }
+  }
+  return totalSize;
+}
+
+std::shared_ptr<arrow::RecordBatch> makeCompressedRecordBatch(
+    uint32_t numRows,
+    const std::vector<std::shared_ptr<arrow::Buffer>>& buffers,
+    const std::shared_ptr<arrow::Schema> compressWriteSchema,
+    ShuffleBufferPool* pool,
+    arrow::util::Codec* codec) {
+  std::vector<std::shared_ptr<arrow::Array>> arrays;
+  // header col, numRows
+  {
+    std::shared_ptr<arrow::Buffer> headerBuffer;
+    GLUTEN_THROW_NOT_OK(pool->allocate(headerBuffer, sizeof(uint32_t)));
+    memcpy(headerBuffer->mutable_data(), &numRows, sizeof(uint32_t));
+    int32_t compressType = static_cast<int32_t>(codec->compression_type());
+    memcpy(headerBuffer->mutable_data() + sizeof(uint32_t), &compressType, sizeof(int32_t));
+    arrays.emplace_back(makeBinaryArray(compressWriteSchema->field(0)->type(), headerBuffer, pool));
+  }
+
+  // Length buffer layout |buffers.size()|buffer1 unCompressedLength|buffer1 compressedLength| buffer2...
+  std::shared_ptr<arrow::Buffer> lengthBuffer;
+  GLUTEN_THROW_NOT_OK(pool->allocate(lengthBuffer, (buffers.size() * 2 + 1) * sizeof(int64_t)));
+  int64_t offset = 0;
+  writeInt64(lengthBuffer, offset, buffers.size());
+
+  int64_t compressedBufferMaxSize = getMaxCompressedBufferSize(buffers, codec);
+  std::shared_ptr<arrow::ResizableBuffer> valueBuffer;
+  GLUTEN_THROW_NOT_OK(pool->allocateDirectly(valueBuffer, compressedBufferMaxSize));
+  int64_t compressValueOffset = 0;
+  for (auto& buffer : buffers) {
+    if (buffer != nullptr && buffer->size() != 0) {
+      writeInt64(lengthBuffer, offset, buffer->size());
+      int64_t actualLength;
+      int64_t maxLength = codec->MaxCompressedLen(buffer->size(), nullptr);
+      GLUTEN_ASSIGN_OR_THROW(
+          actualLength,
+          codec->Compress(
```

Comment:
We should give up using arrow codec and switch to the original compression API, like `LZ4F_compressFrameBound`. I have a demo here https://github.com/marin-ma/gluten/commit/c81b99f9927f85b628947947efc1c61c4f911e25#diff-49b33dd9e5aaf7f7eb6e475fe13595eac27e5a9b978590b3cee815c591ab9e47R76-R195

Reply:
Also worth to try zstd streaming API to compress multiple buffers into one.

Reply:
> We should give up using arrow codec and switch to the original compression API, like `LZ4F_compressFrameBound`. I have a demo here [marin-ma@c81b99f#diff-49b33dd9e5aaf7f7eb6e475fe13595eac27e5a9b978590b3cee815c591ab9e47R76-R195](https://github.com/marin-ma/gluten/commit/c81b99f9927f85b628947947efc1c61c4f911e25#diff-49b33dd9e5aaf7f7eb6e475fe13595eac27e5a9b978590b3cee815c591ab9e47R76-R195)

Here you copy all the buffers into a big buffer, and then compress this big buffer, the usage is different.

Reply:
Can you please add a link to zstd streaming API?

Reply:
https://github.com/marin-ma/gluten/commit/c81b99f9927f85b628947947efc1c61c4f911e25#diff-49b33dd9e5aaf7f7eb6e475fe13595eac27e5a9b978590b3cee815c591ab9e47R133-R178 This functions doesn't copy buffers, and use zstd streaming API.